src/speculators/data_generation/preprocessing.py

```diff
@@ -0,0 +1,262 @@
+    if chat_template not in CHAT_TEMPLATES:
+        raise ValueError(
+            f"Chat template '{chat_template}' not found. "
+            f"Available templates: {list(CHAT_TEMPLATES.keys())}"
+        )
+    template = CHAT_TEMPLATES[chat_template]
+    original_cols = dataset.column_names
+
+    dataset = dataset.map(
+        lambda examples: _preprocess_batch(examples, tokenizer, template, max_length),
+        batched=True,
+        num_proc=num_proc,
+        batch_size=1000,
+        remove_columns=original_cols,
+        load_from_cache_file=False,
+    )
+
+    dataset.set_format(type="torch")
+    return dataset
+
+
+def load_raw_dataset(train_data_path: str, num_proc: int = 8) -> HFDataset:
+    """Load raw dataset from local file or HuggingFace."""
+    if train_data_path.endswith((".jsonl", ".json")):
+        return load_dataset("json", data_files=train_data_path, split="train")
+
+    if train_data_path not in DATASET_CONFIGS:
+        raise ValueError(
+            f"Unsupported dataset: {train_data_path}. "
+            f"Supported: local .json/.jsonl files or {list(DATASET_CONFIGS.keys())}"
+        )
+
+    config = DATASET_CONFIGS[train_data_path]
+    raw_dataset = load_dataset(config.hf_path, split=config.split)
+
+    if config.normalize_fn is not None:
+        raw_dataset = raw_dataset.map(config.normalize_fn, num_proc=num_proc)
+
+    return raw_dataset
+
+
+def generate_cache_key(
+    target_model_path: str,
+    chat_template: str,
+    seq_length: int,
+    train_data_path: str,
+) -> str:
+    """Generate MD5 cache key from preprocessing parameters."""
+    key_string = f"{target_model_path}_{chat_template}_{seq_length}_{train_data_path}"
+    return hashlib.md5(key_string.encode()).hexdigest()
+
+
+def load_and_preprocess_dataset(
+    target_model_path: str,
+    train_data_path: str,
+    chat_template: str,
+    seq_length: int,
+    cache_dir: str,
+    build_dataset_num_proc: int = 8,
+    seed: int = 0,
+    max_samples: int | None = None,
+) -> tuple[HFDataset, PreTrainedTokenizer]:
+    """Load, tokenize, and preprocess a dataset for EAGLE3 training."""
+    log.section("Starting dataset preprocessing")
+
+    if chat_template not in CHAT_TEMPLATES:
+        raise ValueError(
+            f"Chat template '{chat_template}' not found. "
+            f"Available: {list(CHAT_TEMPLATES.keys())}"
+        )
+
+    log.subsection("Loading tokenizer and dataset")
+    tokenizer = AutoTokenizer.from_pretrained(target_model_path, trust_remote_code=True)
+    if tokenizer.pad_token is None:
+        tokenizer.pad_token = tokenizer.eos_token
+
+    raw_dataset = load_raw_dataset(train_data_path, num_proc=build_dataset_num_proc)
+    raw_dataset = raw_dataset.shuffle(seed=seed)
+
+    if max_samples is not None and len(raw_dataset) > max_samples:
+        raw_dataset = raw_dataset.select(range(max_samples))
+
+    log.info(f"Loaded {len(raw_dataset)} samples")
+
+    cache_key = generate_cache_key(
```

Comment:
FYI some of this functionality is already built into HF Datasets
https://huggingface.co/docs/datasets/en/about_cache